File: Python/array/spiral_matrix.py
from typing import *

def spiralOrder(matrix: List[List[int]]) -> List[int]:
    result = []
    row, col, direct = 0, 0, 0
    dir = [[0, 1], [1, 0], [0, -1], [-1, 0]]
    m, n = len(matrix), len(matrix[0])
    for i in range(m * n):
        result.append(matrix[row][col])
        matrix[row][col] = 1000
        next_row, next_col = row + dir[direct][0], col + dir[direct][1]
        if next_row < 0 or next_row >= m or next_col < 0 or next_col >= n or matrix[next_row][next_col] == 1000:
            direct = (direct + 1) % 4
        row, col = row + dir[direct][0], col + dir[direct][1]
    
    return result
        

File: core/tmp/Java/src/com/dbc/algorithm/Array/SpiralMatrix.java
package com.dbc.algorithm.Array;

import java.util.ArrayList;
import java.util.List;

public class SpiralMatrix {
    public List<Integer> spiralOrder(int[][] matrix) {
        int[][] direct = new int[][]{{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        int m = matrix.length, n = matrix[0].length;
        int row = 0, col = 0, dir = 0;
        List<Integer> result = new ArrayList<>();
        for (int i = 0; i < m * n; i++) {
            result.add(matrix[row][col]);
            matrix[row][col] = 1000;
            int nextRow = row + direct[dir][0], nextCol = col + direct[dir][1];
            if (nextRow < 0 || nextRow >= m || nextCol < 0 || nextCol >= n || matrix[nextRow][nextCol] == 1000) {
                dir = (dir + 1) % 4;
            }
            row += direct[dir][0];
            col += direct[dir][1];
        }
        return result;
    }
}
